### PR TITLE
6hundreds/diff util adapter

### PR DESCRIPTION
**Pull request by 6hundreds**
Hi Hannes! 
First of all - thank you for such elegant solution of multiple viewTypes in RecyclerView.
In this PR i'm implemented new adapter type, that allows to work with ListAdapter from support library 27.0.1. It using AsyncListDiffer internally and calculating diffs in the background thread by the default.
New sample was added
Readme was updated
Tests for new classes written

### CODE REVIEW

File: library/src/test/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapterTest.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Test
+    public void checkDiffCallbackInstance() {
+        final DiffCallbackImpl<DiffItem> callback = new DiffCallbackImpl<>();
+        new DiffDelegationAdapter<DiffItem>(callback) {
+            @Override
+            public int getItemCount() {
+                AsyncListDiffer asyncListDiffer = new AsyncListDiffer<>(this, callback);
+                Assert.assertEquals(differ, asyncListDiffer);
```
Comment (sockeqwe) [inline library/src/test/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapterTest.java]: what are you trying to test here? That `callback` has been set properly inside `DiffDelegationAdapter`?

File: library/src/test/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapterTest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+/**
+ * Created by six_hundreds on 28.06.18.
+ */
+@RunWith(RobolectricTestRunner.class)
```
Comment (sockeqwe) [inline library/src/test/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapterTest.java]: If possible I would like to keep Robolectirc out of this library. From experience I know it adds just some extra maintainance costs and overall it looks like it adds very little value to have the following tests compared to the maintainance cost. What do you think?

File: library/build.gradle [hunk 3]
```diff
@@ -26,6 +26,7 @@ dependencies {
   api rootProject.ext.recyclerView
 
   testImplementation 'junit:junit:4.12'
+  testImplementation 'org.robolectric:robolectric:3.8'
```
Comment (sockeqwe) [inline library/build.gradle]: Why do we need Robolectiric? Is there no other way around this? I would like to keep Robolectric out of this library. It causes additional maintainance costs and often it indicates some API flaws that we might could solve properly in a different way.

File: README.md [hunk 4]
```diff
@@ -184,6 +184,54 @@ public class CatListItemAdapterDelegate extends AbsListItemAdapterDelegate<Cat,
...
+If you are using Kotlin you are actually have nothing to do. Kotlin `data class` already has `hashCode()` implementation. It case of Java you will need to generate `equals()` and `hashCode()` methods.
+
+You have a `DiffDelegationAdapter` which is ready to go implementation of `AbsDiffDelegationAdapter` with API rather similar to `ListDelegationAdapter`. 
+
+```java
+public class DiffAdapter extends DiffDelegationAdapter<DiffItem> {
+
+    public DiffAdapter() {
+        delegatesManager.addDelegate(new DiffDogAdapterDelegate());
+        delegatesManager.addDelegate(new DiffCatAdapterDelegate());
```
Comment (sockeqwe) [inline README.md]: Can be simplified to 
```java
delegatesManager.addDelegate(new DiffDogAdapterDelegate())
                .addDelegate(new DiffCatAdapterDelegate());
```

File: README.md [hunk 5]
```diff
@@ -184,6 +184,54 @@ public class CatListItemAdapterDelegate extends AbsListItemAdapterDelegate<Cat,
...
+    }
+}
+```
+
+And provide data to it through `setItems()` method. No more carry about which items was deleted, inserted or etc. `AsyncListDiffer` will do all the job for you and will apply all particular animations to your list.
```
Comment (sockeqwe) [inline README.md]: I would rephrase this a bit: 

You must set data by calling `setItems()`. Internally `AsyncListDiffer` computes and applies all item animations for you.

What do you think about that?

File: README.md [hunk 6]
```diff
@@ -184,6 +184,54 @@ public class CatListItemAdapterDelegate extends AbsListItemAdapterDelegate<Cat,
 
 As you see, instead of writing code that casts list item to Cat we can use `AbsListItemAdapterDelegate` to do the same job (by declaring generic types).
 
+## DiffUtil & ListAdapter
```
Comment (sockeqwe) [inline README.md]: Awesome documenation. Well done 👍

File: app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/adapterdelegates/DiffCatAdapterDelegate.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import com.hannesdorfmann.adapterdelegates.sample.R;
+import com.hannesdorfmann.adapterdelegates3.AbsListItemAdapterDelegate;
+import com.hannesdorfmann.adapterdelegates3.DiffItem;
+import com.hannesdorfmann.adapterdelegates3.sample.model.DiffCat;
+
+import java.util.List;
+
+/**
+ * Created by six_hundreds on 15.06.18.
+ */
```
Comment (sockeqwe) [inline app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/adapterdelegates/DiffCatAdapterDelegate.java]: Please remove this comment or make it more meaningful (I think removing it is better).

File: app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/adapterdelegates/DiffDogAdapterDelegate.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import com.hannesdorfmann.adapterdelegates.sample.R;
+import com.hannesdorfmann.adapterdelegates3.AbsListItemAdapterDelegate;
+import com.hannesdorfmann.adapterdelegates3.DiffItem;
+import com.hannesdorfmann.adapterdelegates3.sample.model.DiffDog;
+
+import java.util.List;
+
+/**
+ * Created by six_hundreds on 15.06.18.
+ */
```
Comment (sockeqwe) [inline app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/adapterdelegates/DiffDogAdapterDelegate.java]: Please remove this comment or make it more meaningful (I think removing it is better).

File: app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/model/DiffCat.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
+package com.hannesdorfmann.adapterdelegates3.sample.model;
+
+import com.hannesdorfmann.adapterdelegates3.DiffItem;
+
+/**
+ * Created by six_hundreds on 15.06.18.
+ */
```
Comment (sockeqwe) [inline app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/model/DiffCat.java]: Please remove this comment or make it more meaningful (I think removing it is better).

File: app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/model/DiffDog.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
+package com.hannesdorfmann.adapterdelegates3.sample.model;
+
+import com.hannesdorfmann.adapterdelegates3.DiffItem;
+
+/**
+ * Created by six_hundreds on 15.06.18.
+ */
```
Comment (sockeqwe) [inline app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/model/DiffDog.java]: Please remove this comment or make it more meaningful (I think removing it is better).

File: library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapter.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+ * @author Sergey Opivalov
+ */
+
+public class DiffDelegationAdapter<T extends DiffItem> extends AbsDiffDelegationAdapter<T> {
+
+    public DiffDelegationAdapter() {
+        super(new DiffCallbackImpl<T>());
+    }
+
+    public DiffDelegationAdapter(@NonNull DiffUtil.ItemCallback<T> itemCallback) {
```
Comment (sockeqwe) [inline library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapter.java]: If you can pass in any arbitarry `DiffUtil.ItemCallback<T>` then the restriction that it must be `DiffItem` is not useful anymore because one could write a `DiffUtil.ItemCallback` that is not using `DiffItem.getItemId()` nor `DiffItem.getItemHash()`.

As I also pointed out in some other comments, I'm not sure if this library should contain something like `DiffItem` and a default `DiffCallbackImpl` respectively.

File: library/src/main/java/com/hannesdorfmann/adapterdelegates3/AbsDiffDelegationAdapter.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+ */
+
+public abstract class AbsDiffDelegationAdapter<T extends DiffItem> extends RecyclerView.Adapter {
+
+    protected final AdapterDelegatesManager<List<T>> delegatesManager;
+    protected final AsyncListDiffer<T> differ;
+
+    public AbsDiffDelegationAdapter(@NonNull DiffUtil.ItemCallback<T> diffCallback) {
+        this.differ = new AsyncListDiffer<>(this, diffCallback);
+        this.delegatesManager = new AdapterDelegatesManager<>();
```
Comment (sockeqwe) [inline library/src/main/java/com/hannesdorfmann/adapterdelegates3/AbsDiffDelegationAdapter.java]: You should call other (primary) constructor `this(diffCallback, new AdapterDelegatesManager<>())` instead of duplicating code.

File: app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/DiffActivity.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+import com.hannesdorfmann.adapterdelegates3.sample.model.DiffCat;
+import com.hannesdorfmann.adapterdelegates3.sample.model.DiffDog;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Created by six_hundreds on 15.06.18.
+ */
```
Comment (sockeqwe) [inline app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/DiffActivity.java]: Please remove this comment or make it more meaningful (I think removing it is better).

File: library/src/main/java/com/hannesdorfmann/adapterdelegates3/AbsDiffDelegationAdapter.java [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    public AbsDiffDelegationAdapter(@NonNull DiffUtil.ItemCallback<T> diffCallback,
```
Comment (sockeqwe) [inline library/src/main/java/com/hannesdorfmann/adapterdelegates3/AbsDiffDelegationAdapter.java]: If you can pass in any arbitarry DiffUtil.ItemCallback<T> then the restriction that it must be DiffItem is not useful anymore because one could write a DiffUtil.ItemCallback that is not using DiffItem.getItemId() nor DiffItem.getItemHash().

As I also pointed out in some other comments, I'm not sure if this library should contain something like DiffItem and a default DiffCallbackImpl respectively.

File: library/src/test/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapterTest.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        };
+
+    }
+
+    @Test
+    public void checkDiffCallbackNotNull() {
+        new DiffDelegationAdapter() {
+            @Override
+            public int getItemCount() {
+                Assert.assertNotNull(differ);
```
Comment (sockeqwe) [inline library/src/test/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapterTest.java]: This case is alread checked in test case above (because it implies that its not null)

File: library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffCallbackImpl.java [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+public class DiffCallbackImpl<T extends DiffItem> extends DiffUtil.ItemCallback<T> {
+
+    @Override
+    public boolean areItemsTheSame(T oldItem, T newItem) {
+        return oldItem.getItemId() == newItem.getItemId();
+    }
+
+    @Override
+    public boolean areContentsTheSame(T oldItem, T newItem) {
+        return oldItem.getItemHash() == newItem.getItemHash();
```
[the remark this answers is not in the document]
Comment (6hundreds) [reply library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffCallbackImpl.java]: I'm completely agree with you, but AbsDiffDelegationAdapter gets DiffUtil.ItemCallback as a constructor argument. So actually we are let the users to choose which implementation should be used. Isn't it? What do you think? DiffCallbackImpl it's just a default implementation, that can be used in most cases

File: app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/DiffAdapter.java [hunk 17]
```diff
@@ -0,0 +1,18 @@
+package com.hannesdorfmann.adapterdelegates3.sample;
+
+import com.hannesdorfmann.adapterdelegates3.DiffDelegationAdapter;
+import com.hannesdorfmann.adapterdelegates3.DiffItem;
+import com.hannesdorfmann.adapterdelegates3.sample.adapterdelegates.DiffCatAdapterDelegate;
+import com.hannesdorfmann.adapterdelegates3.sample.adapterdelegates.DiffDogAdapterDelegate;
+
+/**
+ * Created by six_hundreds on 15.06.18.
+ */
```
[the remark this answers is not in the document]
Comment (6hundreds) [reply app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/DiffAdapter.java]: Sure

File: library/src/main/java/com/hannesdorfmann/adapterdelegates3/AbsDiffDelegationAdapter.java [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+ *            this.delegatesManager.add(new FooAdapterDelegate());
+ *            this.delegatesManager.add(new BarAdapterDelegate());
+ *        }
+ *    }
+ * }
+ * </pre>
+ *
+ * @param <T> The type of the item, must implement {@link DiffItem}
+ * @author Sergey Opivalov
+ */
```
[the remark this answers is not in the document]
Comment (6hundreds) [reply library/src/main/java/com/hannesdorfmann/adapterdelegates3/AbsDiffDelegationAdapter.java]: Sure

File: library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffItem.java [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+ * }
+ *
+ * @author Sergey Opivalov
+ */
+public interface DiffItem {
+
+    long getItemId();
+
+    int getItemHash();
+}
```
[the remark this answers is not in the document]
Comment (6hundreds) [reply library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffItem.java]: Yes, at the first look approach with hashes and ids does't look like flexible enough. But what do you think about next idea: 
DiffItem will has next definition - 
```
interface DiffItem<T extends DiffItem<T>> {
    boolean areItemsTheSame(T item);
    boolean areContentTheSame(T item);
}
```
In this case item will be responsible itself for comparing with another one. And then we can use hashCodes, unique IDs, etc. That's most common and flexible approach in my opinion. What do you think about this?

File: library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapter.java (hunk 11 above)
Comment (6hundreds) [reply library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapter.java]: But T is restricted with DiffItem interface in DiffDelegationAdapter class signature.

File: library/src/main/java/com/hannesdorfmann/adapterdelegates3/AbsDiffDelegationAdapter.java (hunk 12 above)
Comment (6hundreds) [reply library/src/main/java/com/hannesdorfmann/adapterdelegates3/AbsDiffDelegationAdapter.java]: Sure

File: app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/DiffActivity.java (hunk 13 above)
Comment (6hundreds) [reply app/src/main/java/com/hannesdorfmann/adapterdelegates3/sample/DiffActivity.java]: Sure

File: library/src/main/java/com/hannesdorfmann/adapterdelegates3/AbsDiffDelegationAdapter.java (hunk 14 above)
Comment (6hundreds) [reply library/src/main/java/com/hannesdorfmann/adapterdelegates3/AbsDiffDelegationAdapter.java]: But T is restricted with DiffItem interface in DiffDelegationAdapter class signature.

File: library/src/test/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapterTest.java (hunk 15 above)
Comment (6hundreds) [reply library/src/test/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapterTest.java]: Ok

File: library/src/test/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapterTest.java (hunk 2 above)
Comment (6hundreds) [reply library/src/test/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapterTest.java]: Yeah, i know about that costs. But ListAdapter internally uses Looper, which can't be mocked without Robolectric.

File: README.md (hunk 5 above)
Comment (6hundreds) [reply README.md]: Yeah, i'm agree

File: library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapter.java (hunk 11 above)
Comment (sockeqwe) [reply library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapter.java]: but no one is forced to actually use `DiffItem` api's like `DiffItem. getItemHash()`. The point I try to make is that you could pass in a DiffCallback that is not using `DiffItem` api at all, hence this restriction to T extends DiffItems gives guarantees. Since there are no guarantees, the design is flawn a bit. Thats the point I try to make.

Comment (6hundreds) [reply library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapter.java]: Yes, i've got it. So do you have any ideas about it?

Comment (6hundreds) [reply library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapter.java]: But, again, DiffItem has only two methods,so you'll not be able to implement DiffUtil.Callback with those restriction if you will not use DiffItem api, isn't it?

Comment (sockeqwe) [reply library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapter.java]: Thanks for your Feedback. Maybe I'm too nitpicking but I still think that `T extends DiffItem` adds no guarantees and therefore I would like to discuss if it makes sense to have that at all.

Example how you can trick `T extends DiffItem` restriction:

```java
class Foo implements DiffItem {
    public int id;
    public String name;

     @Override public  long getItemId() { return 0; } // Return value Doesn't really matter because we don't use it anyway

     @Override int getItemHash() { return 0; } // Return value Doesn't really matter because we don't use it anyway
}
```

```java
class MyCallback extends DiffUtil.ItemCallback<Foo> {
    @Override public boolean areItemsTheSame(Foo a,  Foo b){
         return a.name.length == b.name.length;
    }
    boolean areContentTheSame(T item) {
        return a.name.equals(b.name);
    }
}
```

```java
DiffDelegationAdapter<Foo> adapter = new DiffDelegationAdapter<>( MyCallback() );
```

This compiles just fine but it actually contradicts the public API contract since `Foo.getItemId()` nor `Foo. getItemHash()` is used at all although the public API contract of `DiffItem` specifies that this two methods must be used.

That's the point I try to make. A public API should not allow this. So there are two possibility to fix this:

1. Don't allow people to pass in a custom `DiffUtil.ItemCallback`. This means that always the internally `DiffCallbackImpl`is used. By doing so, the restriction `T extends DiffItem` holds because there is no way around properly implementing `DiffItem.getItemId()` and `DiffItem.getItemHash()`  since AdapterDelegates library calls this methods for sure.
2. Remove `T extends DiffItem` as this condition adds no guarantees to the public API contract as shown in my code snippets above.  Furthermore, this means that `DiffItem` is useless at all and should removed too.

I tend to say we should go with route number 2 as I, personally, think that option number 1 is to restritive. Making AdapterDelegates make this decision for the user of the library is an necessary move in my oppinion.  I do understand you motivation to even more simplify things and remove boilerplate but I think the `DiffUtil.ItemCallback` is already simple enough so that whoever wants to use `DiffDelegationAdapter` is able to implement a `DiffUtil.ItemCallback`.

With that being said, nobody actually stops anyone defining something like `DiffItem` in your app and use `DiffItem` and the corresponding `DiffCallbackImpl` in your app. I just wouldn't put it into AdapterDelegates library.

What do you think?

Comment (6hundreds) [reply library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapter.java]: So, i do understand your thoughts. When i took a first look at AdapterDelegates, i was wondered why there is no some interface marker like DisplayableItem or smth like that =) But now i'm understanding your approach =) So, actually i'm agree with you, let's give to users possibility to implement DiffItem by themselves. But how do you think, maybe we can make some example in README?

Comment (sockeqwe) [reply library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapter.java]: Yeah, that makes a ton of sense to me. Let's do it.

File: library/src/test/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapterTest.java (hunk 2 above)
Comment (6hundreds) [reply library/src/test/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapterTest.java]: I'm done with fixes. So what about tests? We can move out Robolectric, but that's mean that we will not be able to work with ListAdapter and basically not be able to test it somehow

**Comment by sockeqwe** [comment]
Thanks for this.
I'm on holidays until Wednesday next week and reviewing code on mobile is not fun :)

Regarding testing, well, if there is really no other way than we might use robolectric, but I would like to take a look at Mockito. I think you can create a new instance with Mockito (its in the constructor where the looper ia set, right) and then use the mocked instance to call onCreateViewHolder() etc. on it. With Movkito you can say that the real methods (not mocked methods) of a movked object should be called. I think this could work in our case.

**Comment by sockeqwe** [comment]
Thanks @6hundreds !
I will take a look at the missing pieces in testing and merge it finally to master soon.
